lib/logstash/plugin.rb

```diff
@@ -10,119 +12,115 @@ class LogStash::Plugin
 
   NL = "\n"
 
-  public
   def hash
-    params.hash ^
-    self.class.name.hash
+    params.hash ^ self.class.name.hash
   end
 
-  public
   def eql?(other)
     self.class.name == other.class.name && @params == other.params
   end
 
-  public
   def initialize(params=nil)
     @params = params
     @logger = Cabin::Channel.get(LogStash)
+    @stop_called = Concurrent::AtomicBoolean.new(false)
   end
 
-  # This method is called when someone or something wants this plugin to shut
-  # down. When you successfully shutdown, you must call 'finished'
-  # You must also call 'super' in any subclasses.
-  public
-  def shutdown(queue)
-    # By default, shutdown is assumed a no-op for all plugins.
-    # If you need to take special efforts to shutdown (like waiting for
-    # an operation to complete, etc)
-    teardown
-    @logger.info("Received shutdown signal", :plugin => self)
-
-    @shutdown_queue = queue
-    if @plugin_state == :finished
-      finished
-    else
-      @plugin_state = :terminating
-    end
-  end # def shutdown
-
-  # You should call this method when you (the plugin) are done with work
-  # forever.
-  public
-  def finished
-    # TODO(sissel): I'm not sure what I had planned for this shutdown_queue
-    # thing
-    if @shutdown_queue
-      @logger.info("Sending shutdown event to agent queue", :plugin => self)
-      @shutdown_queue << self
-    end
-
-    if @plugin_state != :finished
-      @logger.info("Plugin is finished", :plugin => self)
-      @plugin_state = :finished
-    end
-  end # def finished
-
-  # Subclasses should implement this teardown method if you need to perform any
-  # special tasks during shutdown (like flushing, etc.)
-  public
-  def teardown
-    # nothing by default
-    finished
+  # if you override stop, don't forgt to call super
+  def stop
+    @logger.debug("stopping", :plugin => self)
+    @stop_called.make_true
   end
 
-  # This method is called when a SIGHUP triggers a reload operation
-  public
-  def reload
-    # Do nothing by default
+  # stop? should never be overriden
+  def stop?
+    @stop_called.value
   end
 
-  public
-  def finished?
-    return @plugin_state == :finished
-  end # def finished?
-
-  public
-  def running?
-    return @plugin_state != :finished
-  end # def finished?
-
-  public
-  def terminating?
-    return @plugin_state == :terminating
-  end # def terminating?
-
-  public
-  def to_s
-    return "#{self.class.name}: #{@params}"
+  # if you override close, don't forgt to call super
+  def close
```

Comment:
I know we've had this conversation but can this method be named teardown so all cleanup methods are consistently named across plugin types (filters and outputs still use teardown)?
My two reasons for making this teardown are:

1. less changes needed: otherwise we need to refactor all plugins that define `teardown` to use `close` instead
2. arguably a better metaphor: `close` might make sense in plugins such as tcp/udp where you cleanup by closing a socket, but in others the plugin isn't actually closing anything. Therefore `teardown` ([pronunciation](https://www.youtube.com/watch?v=oQAJAYj-s3M)) is more suitable, used commonly in test suites to convey the "cleanup" action.